src/Futhark/CLI/Eval.hs

```diff
@@ -0,0 +1,155 @@
+module Futhark.CLI.Eval (main) where
+  
+import Control.Exception
+import Control.Monad
+import Control.Monad.Except
+import Control.Monad.Free.Church
+
+import Data.Maybe
+import Data.Text qualified as T
+
+import Futhark.Compiler
+import Futhark.MonadFreshNames
+-- import Futhark.Util (toPOSIX)
+import Futhark.Util.Options
+
+import Language.Futhark
+import Language.Futhark.Interpreter qualified as I
+import Language.Futhark.Parser
+import Language.Futhark.Semantic qualified as T
+import Language.Futhark.TypeChecker qualified as T
+
+import System.FilePath
+
+import Futhark.Pipeline
+
+import System.IO
+import Prelude
+-- import Language.Futhark.Interpreter (interpretExp, ExtOp)
+
+
+import Futhark.Util.Pretty
+import Data.List (foldl')
+
+main :: String -> [String] -> IO ()
+-- main _ _ = putStrLn "Hello World"
+main = mainWithOptions interpreterConfig options "options... <exprs...>" run
+  where
+    run [] _ = Nothing
+    run exprs config = Just $ runExprs exprs config
+
+runExprs :: [String] -> InterpreterConfig -> IO ()
+runExprs exprs cfg = do
+  -- putStrLn "Running expressions"
+  let InterpreterConfig _ file = cfg
+  -- putStrLn "File has come up."
+  maybe_new_state <- newFutharkiState cfg file
+  -- putStrLn "New state made"
+  (src, env, ctx) <- case maybe_new_state of
+    Left _ -> error "Unable to load file."
+    Right s -> pure s
+  -- putStrLn "Starting on the expressions"
+  foldl' (\_ b -> runExpr src env ctx b) (putStr "") exprs
+  putStrLn ""
+-- runExprs _ _ = putStrLn "Running Exprs"
+
+
+
+-- Use parseExp, checkExp, then interpretExp.
+runExpr :: VNameSource -> T.Env -> I.Ctx -> String -> IO ()
+runExpr src env ctx str = do
+  uexp <- case parseExp "" (T.pack str) of 
+    Left _ -> error "Syntax Error"
+    Right e -> pure e
+  let (_, expdat) = T.checkExp [] src env uexp
+  fexp <- case expdat of
```

Comment:
Do this directly in the `case`: `case TC.checkExp [] src env uexp of (_, Left _) -> ...`

Reply:
done.